### PR TITLE
Created Hatch marker styles Demo for Example Gallery

### PR DESCRIPTION
**Pull request by forrestm**
## PR Summary
Screen Shot 2020-07-10 at 6 07 33 PM

This is a hatch marker styles demo. The script will plot 10 squares filled with 10 hatching markers. There was no example that demonstrated the 10 different hatching markers. Thanks.

## PR Checklist

- [ ] Has Pytest style unit tests
- [ ] Code is [Flake 8](http://flake8.pycqa.org/en/latest/) compliant
- [ ] New features are documented, with examples if plot related
- [ ] Documentation is sphinx and numpydoc compliant
- [ ] Added an entry to doc/users/next_whats_new/ if major new feature (follow instructions in README.rst there)
- [ ] Documented in doc/api/next_api_changes/* if API changed in a backward-incompatible way

### CODE REVIEW

File: examples/shapes_and_collections/hatches_styles.py [hunk 1]
```diff
@@ -0,0 +1,47 @@
...
+import matplotlib.pyplot as plt
+import numpy as np
+from matplotlib.patches import Rectangle
+
+fig, ax = plt.subplots()
+
+x = np.repeat([1, 4, 7, 10, 13], 2)
+y = np.tile([2, 5], 5)
+
+pos = np.array(list(zip(x, y)))
```
Comment (QuLogic) [inline examples/shapes_and_collections/hatches_styles.py:18]:
```suggestion
pos = np.column_stack((x, y))
```

File: examples/shapes_and_collections/hatches_styles.py [hunk 2]
```diff
@@ -0,0 +1,47 @@
...
+text_pos = pos + [0.9, -0.5]
+
+hashes = ['/', '\\', '|', '-', '+', 'x', 'o', 'O', '.', '*']
+
+for h in range(len(hashes)):
+    ax.add_patch(Rectangle(pos[h], 2, 2, fill=False, hatch=hashes[h]))
+    ax.text(text_pos[h][0], text_pos[h][1], hashes[h], fontsize=15)
```
Comment (QuLogic) [inline examples/shapes_and_collections/hatches_styles.py:25]:
```suggestion
for h, p, t in zip(hashes, pos, text_pos):
    ax.add_patch(Rectangle(p, 2, 2, fill=False, hatch=h))
    ax.text(t[0], t[1], h, fontsize=15)
```

File: examples/shapes_and_collections/hatches_styles.py [hunk 3]
```diff
@@ -0,0 +1,47 @@
...
+
+plt.axis('equal')
+plt.axis('off')
+plt.tight_layout()
```
Comment (QuLogic) [inline examples/shapes_and_collections/hatches_styles.py:29]:
```suggestion
ax.axis('equal')
ax.axis('off')
fig.tight_layout()
```

**Comment by QuLogic** [comment]
You'll need to fix the doc build:
```
/home/circleci/project/doc/gallery/shapes_and_collections/hatches_styles.rst:12: WARNING: Title overline too short.

==========
Hatches Styles Demo
==========
```
and flake8, by adding an entry to `.flake8`.

File: examples/shapes_and_collections/hatches_styles.py [hunk 4]
```diff
@@ -0,0 +1,47 @@
...
+
+hashes = ['/', '\\', '|', '-', '+', 'x', 'o', 'O', '.', '*']
+
+for h, p, t in zip(hashes, pos, text_pos):
+    ax.add_patch(Rectangle(p, 2, 2, fill=False, hatch=h))
+    ax.text(t[0], t[1], h, fontsize=15)
+
+ax.axis('equal')
+ax.axis('off')
+fig.tight_layout()
```
Comment (story645) [inline examples/shapes_and_collections/hatches_styles.py:29]: will `plt.subplots(constrained_layout=True)` work here?

File: examples/shapes_and_collections/hatches_styles.py [hunk 5]
```diff
@@ -0,0 +1,47 @@
...
+y = np.tile([2, 5], 5)
+
+pos = np.column_stack((x, y))
+text_pos = pos + [0.9, -0.5]
+
+hashes = ['/', '\\', '|', '-', '+', 'x', 'o', 'O', '.', '*']
+
+for h, p, t in zip(hashes, pos, text_pos):
+    ax.add_patch(Rectangle(p, 2, 2, fill=False, hatch=h))
+    ax.text(t[0], t[1], h, fontsize=15)
```
Comment (story645) [inline examples/shapes_and_collections/hatches_styles.py:25]: I'd rather that the hatchmark be the title of the patch and possibly it should be `f"'{}'"` - currently it's getting a bit lost as to what exactly it identifies.

File: examples/shapes_and_collections/hatches_styles.py [hunk 6]
```diff
@@ -0,0 +1,47 @@
+"""
+===================
+Hatches Styles Demo
```
Comment (timhoffm) [inline examples/shapes_and_collections/hatches_styles.py:3]:
```suggestion
Hatch style reference
```
- This is rather a reference than a demo.
- Not a native speaker, but I think "Hatches styles" is not grammatically correct.
- Capitalization style: We capitalize only the first letter of a heading (https://matplotlib.org/devdocs/devel/documenting_mpl.html#section-name-formatting).

File: examples/shapes_and_collections/hatches_styles.py [hunk 7]
```diff
@@ -0,0 +1,47 @@
...
+
+fig, ax = plt.subplots()
+
+x = np.repeat([1, 4, 7, 10, 13], 2)
+y = np.tile([2, 5], 5)
+
+pos = np.column_stack((x, y))
+text_pos = pos + [0.9, -0.5]
+
+hashes = ['/', '\\', '|', '-', '+', 'x', 'o', 'O', '.', '*']
```
Comment (timhoffm) [inline examples/shapes_and_collections/hatches_styles.py:21]:
```suggestion
hatches = ['/', '\\', '|', '-', '+', 'x', 'o', 'O', '.', '*']
```

File: examples/shapes_and_collections/hatches_styles.py (hunk 4 above)
Comment (forrestm) [reply examples/shapes_and_collections/hatches_styles.py]: I have added `constrained_layout=True`.

File: examples/shapes_and_collections/hatches_styles.py (hunk 5 above)
Comment (forrestm) [reply examples/shapes_and_collections/hatches_styles.py]: I added `ax.set_title(f"' {h} '")`.

File: examples/shapes_and_collections/hatches_styles.py (hunk 6 above)
Comment (forrestm) [reply examples/shapes_and_collections/hatches_styles.py]: Changed filename to hatch_style_reference.py and updated the Heading.

File: examples/shapes_and_collections/hatches_styles.py (hunk 7 above)
Comment (forrestm) [reply examples/shapes_and_collections/hatches_styles.py]: I corrected this spelling error.

File: examples/shapes_and_collections/hatch_style_reference.py [hunk 8]
```diff
@@ -0,0 +1,40 @@
...
+from matplotlib.patches import Rectangle
+
+fig, axs = plt.subplots(2, 5, constrained_layout=True, figsize=(10, 5))
+
+hatches = ['/', '\\', '|', '-', '+', 'x', 'o', 'O', '.', '*']
+
+
+def hatches_plot(ax, h):
+    ax.add_patch(Rectangle((0, 0), 2, 2, fill=False, hatch=h))
+    ax.set_title(f"' {h} '")
```
Comment (story645) [inline examples/shapes_and_collections/hatch_style_reference.py:19]: You were right that it's better the other way (underneath)

**Comment by jklymak** [comment]
This is great, but what's our policy on adding a new demo versus just modifying the existing one?  

I'm confused by the statement about the relevant backends.  Which backends *don't* support hatching?

**Comment by forrestm** [comment]
@jklymak This is no longer a demo. @timhoffm indicated it should be a reference. I am not aware of the policy on modifying demos. Please let me know if this should be added to an existing demo. 

The information about the backends originated from `hatch_demo.py`. I copied that script as a template. I can remove the note if you think it is not necessary.

File: examples/shapes_and_collections/hatch_style_reference.py (hunk 8 above)
Comment (forrestm) [reply examples/shapes_and_collections/hatch_style_reference.py]: I moved the title below the plots. 
download

**Comment by story645** [comment]
@jklymak so the original issue in particular requested a reference doc because the actual hatch options kind of get buried/lost in the demo and are kinda hard to find in the docs and I wanted a pure listing. Long term, I'd love it if all the reference type examples could move out of the gallery into either their own smaller/curated reference gallery or docs like @brunobeltran proposed.

File: examples/shapes_and_collections/hatch_style_reference.py [hunk 9]
```diff
@@ -0,0 +1,40 @@
+"""
+=====================
+Hatch style reference
+=====================
+
+Hatching (pattern filled polygons) is supported currently in the PS,
+PDF, SVG and Agg backends only.
```
Comment (timhoffm) [inline examples/shapes_and_collections/hatch_style_reference.py:7]: Picking up on @jklymak's comment

> Which backends don't support hatching?

```suggestion
Hatching (pattern filled polygons) is currently supported in the backends
PS, PDF, SVG and *Agg. The backends OSX, WX and *Cairo ignore hatching.
```

**Comment by jklymak** [comment]
This example is fine. I just would add it to the existing demo rather than make a whole new file. My understanding was that we thought our gallery had too many entries.

**Comment by story645** [comment]
So I agree our existing gallery is too large, but I think that's sort of a different discussion on taxonomy. This example is documenting something different then the demo -less how to use it, more what are my options-and is more in keeping with the marker and tick reference examples and therefore should be stand alone.

**Comment by jklymak** [comment]
The gallery entry that discusses hatches could have a demo *and* a reference, all on one page.  I don't see the point in a separate page, one of which does not show all the options, and the other of which does not show how the feature is typically used.  I wouldn't get hung up on whether it is a "demo" or a "reference" or that the two things should be kept pristinely apart.  Quite the opposite in fact - most references are better with examples.

**Comment by story645** [comment]
I think we should have more smaller examples then fewer larger ones 'cause the larger ones make it harder to figure out exactly what the example is trying to show.  Maybe if the example is broken up w/ narrative between the reference and the demo, but that's out of scope here. 
I in particular don't like the hatch demo example because I find it so busy that it's hard to figure out what exactly it's doing - I think 'cause it tries to be both reference and demo. And more importantly, since the gallery only allows for one thumbnail per entry the reference should take priority and then the demo would be swallowed anyway.

**Comment by timhoffm** [comment]
The reference answers the question "Which hatches are available", in contrast to the demo ("How do I plot with hatches").

IMHO there is value in having the reference separate as a linkable target without the fuss of an example. Both reference and demo may cross-link if we like.

This is in the spirit of:

- https://matplotlib.org/devdocs/gallery/lines_bars_and_markers/joinstyle.html
- https://matplotlib.org/devdocs/gallery/lines_bars_and_markers/linestyles.html
- https://matplotlib.org/devdocs/gallery/lines_bars_and_markers/marker_reference.html

which all should/will be moved to a dedicated gallery/section for parameter references.

The hatch demo is really bad and should be updated, but that's a separate topic.

**Comment by jklymak** [comment]
> IMHO there is value in having the reference separate as a linkable target without the fuss of an example. Both reference and demo may cross-link if we like.

Again, I don't see what the value is of having them be separate is; linking one page would be more useful to me than flipping between two pages.

**Comment by story645** [comment]
@jklymak the motivating example for this was a geopandas developer who very much knows how to use hatches just wanting a clear list of them. For that use case, the demo gets in the way b/c frankly in the reference examples the code is basically besides the point. 

And for the users who want to know how to use the thing, the reference generating code tends to be a bit overcomplicated and like I don't think it's a great idea to throw sometimes "scary" code at 'em when they're really just trying to sort out which kwarg makes thing. (And yes this is kinda ignoring my and @timhoffm 's consensus that the hatch demo is independently bad at the thing it's supposed to be doing).

**Comment by jklymak** [comment]
So, this PR is fine the way it is.  

@story645 lets just leave it at that I agree with your previous stance that there are too many gallery entries, and disagree with your current stance that we need a different gallery entry for each possible type of user 😛

**Comment by jklymak** [comment]
@forrestm Thanks a lot for the contribution, and sorry to hijack your PR!

**Comment by story645** [comment]
It's for each different type of use 😋

**Comment by forrestm** [comment]
Thanks for letting me work on this. I look forward to working on more issues. Thanks everyone!

**Pull request merged**